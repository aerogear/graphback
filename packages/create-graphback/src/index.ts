#!/usr/bin/env node

import * as yargs from 'yargs';

import { init } from './init/components';

type Params = {
  name?: string;
  templateName?: string;
  templateUrl?: string;
  filter?: string;
};

const command = '$0 <name>';

export const desc = 'Create Graphback project from available templates';

// tslint:disable-next-line: typedef
export const builder = (args: yargs.Argv) => {
  args.positional('name', {
    describe: 'Project name',
    type: 'string'
  });
  args.option('templateName', {
    describe: 'Name of the predefined template',
    type: 'string'
  });
  args.option('templateUrl', {
    describe:
      'GitHub URL of the template. For example (https://github.com/wtrocki/graphql-serve-example#master)',
    type: 'string'
  });
  args.option('filter', {
    describe:
      'A filter option used to limit the number of displayed templates that only matches it. For example "postgres" will only display templates that have a Postgres backed',
    type: 'string'
  });
};

export async function handler({
  name,
  templateName,
  templateUrl,
  filter
}: Params) {
  await init(name, templateName, templateUrl, filter);
}

if (require.main === module) {
  // eslint-disable-next-line no-unused-expressions
  yargs
    .command(command, desc, builder, handler)
    .example(
      '$0 my-awesome-project',
      'creates the "my-awesome-project" in the current working directory'
    )
    .demandCommand(1)
    .strict()
    .recommendCommands()
    .help()
    .alias('h', 'help')
    .version()
    .alias('v', 'version').argv;
}
