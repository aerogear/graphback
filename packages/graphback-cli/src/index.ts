#!/usr/bin/env node

import * as yargs from 'yargs';

if (require.main === module) {
  // eslint-disable-next-line no-unused-expressions
  yargs
    .commandDir('commands')
    .demandCommand(1)
    .strict()
    .recommendCommands()
    .help()
    .alias('h', 'help')
    .version()
    .alias('v', 'version')
    .argv;
}
