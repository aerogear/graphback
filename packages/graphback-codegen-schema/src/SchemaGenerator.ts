import { InputModelTypeContext } from '@graphback/core';
import { generateSchemaString } from './schemaTemplate';
import { buildTargetContext, createCustomSchemaContext, TargetContext } from './targetSchemaContext';

/**
 * Schema formatter that provides ability to wrap schema string with language specific code
 * that can simplify importing and using it within server side code.
 *
 * @see jsSchemaFormatter
 * @see tsSchemaFormatter
 * @see gqlSchemaFormatter
 *
 */
export interface SchemaFormatter {
    /**
     * Transform schema string to new format.
     * Can be used to wrap schema into js or typescript import format that can be added to the file
     *
     * @param schemaString
     */
    format(schemaString: string): string
}


/**
 * Generate schema string based on the input model.
 *
 * Schema is generated by passing type `inputContext` array containing information about all
 * provided types that needs to be included.
 */
export class SchemaGenerator {
    private context: TargetContext
    private inputContext: InputModelTypeContext[]
    private formatter: SchemaFormatter;

    constructor(inputContext: InputModelTypeContext[], formatter?: SchemaFormatter) {
        this.inputContext = inputContext
        this.formatter = formatter;
    }

    /**
     * Generate output schema as string
     */
    public generate() {
        this.context = buildTargetContext(this.inputContext)
        const customContext = createCustomSchemaContext(this.inputContext)
        const schemaString = generateSchemaString(this.context, customContext)
        if (this.formatter) {
            return this.formatter.format(schemaString);
        }

        return schemaString
    }
}
