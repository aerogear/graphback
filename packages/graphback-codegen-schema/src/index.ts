export * from './writer/schemaFormatters'
export * from './SchemaCRUDPlugin'