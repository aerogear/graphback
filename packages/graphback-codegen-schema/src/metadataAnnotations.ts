export const markers = {
    createdAt: "createdAt",
    updatedAt: "updatedAt",
    versioned: "versioned"
}

export const fieldNames = {
    createdAt: "createdAt",
    updatedAt: "updatedAt"
}