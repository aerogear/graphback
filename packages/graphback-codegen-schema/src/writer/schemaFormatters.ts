import { SchemaFormatter } from './SchemaFormatter'

const noteString = 'NOTE: This schema was generated by Graphback and should not be changed manually'

/**
 * Typescript string template that formats schema into common js module that can be imported 
 * easily in server side application
 */
export const tsSchemaFormatter: SchemaFormatter = {
  format: (schemaString: string) => {
    return `
// ${noteString}
export const schemaString = \`
${schemaString}
\`;
`
  }
}

/**
 * JS string template that formats schema into common js module that can be imported 
 * easily in server side application
 */
export const jsSchemaFormatter: SchemaFormatter = {
  format: (schemaString: string) => {
    return `
// ${noteString}
const schemaString = \`
${schemaString}
\`;

module.exports = { schemaString };
`
  }
}


/**
 * GQL string template that returns schema in original form
 */
export const gqlSchemaFormatter: SchemaFormatter = {
  format: (schemaString: string) => {
    return `## ${noteString}\n\n${schemaString}`;
  }
}