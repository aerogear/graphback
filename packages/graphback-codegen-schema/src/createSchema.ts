import { InputModelTypeContext } from '@graphback/core'
import { SchemaGenerator } from './SchemaGenerator';
import { gqlSchemaFormatter, jsSchemaFormatter, tsSchemaFormatter } from './writer/schemaFormatters';

export interface SchemaWriterOptions {
    // output format
    format: 'ts' | 'js' | 'gql'
}

/**
 * Create resolvers function that 
 * @param inputContext 
 * @param options 
 */
export const createSchema = (inputContext: InputModelTypeContext[], options: SchemaWriterOptions): string => {
    if (options) {
        if (options.format === 'ts') {
            return new SchemaGenerator(inputContext, tsSchemaFormatter).generate();
        }
        if (options.format === 'js') {
            return new SchemaGenerator(inputContext, jsSchemaFormatter).generate();;
        }
        if (options.format === 'gql') {
            return new SchemaGenerator(inputContext, gqlSchemaFormatter).generate();
        }
    }
    throw Error("Invalid format specified. `options.format` supports only `ts`, `js` and `gql` flags");
}
