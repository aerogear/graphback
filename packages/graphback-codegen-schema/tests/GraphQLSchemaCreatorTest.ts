import { tsSchemaFormatter, gqlSchemaFormatter } from '../src/schemaFormatters';
// tslint:disable-next-line: match-default-export-name no-implicit-dependencies
import { graphQLInputContext } from '@graphback/core/src';
import ava, { ExecutionContext } from 'ava';
import { readFileSync } from 'fs';
import { buildSchema, parse } from 'graphql';
import { SchemaGenerator } from '../src/SchemaGenerator';

const schemaText = readFileSync(`${__dirname}/mock.graphql`, 'utf8')

ava('Test snapshot config gql', async (t: ExecutionContext) => {
  const defautConfig = {
    "create": true,
    "update": true,
    "findAll": true,
    "find": true,
    "delete": true,
    "subCreate": true,
    "subUpdate": true,
    "subDelete": true,
    "disableGen": true
  }

  const inputContext = graphQLInputContext.createModelContext(schemaText, defautConfig)

  const schemaGenerator = new SchemaGenerator(inputContext, gqlSchemaFormatter)
  const schema = schemaGenerator.generate()
  t.snapshot(schema);
  t.true(parse(schema).definitions.length > 0);
});
