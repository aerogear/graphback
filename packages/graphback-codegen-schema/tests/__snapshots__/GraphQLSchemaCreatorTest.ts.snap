// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test one side relationship schema query type generation 1`] = `
"input IDInput {
  ne: ID
  eq: ID
  le: ID
  lt: ID
  ge: ID
  gt: ID
  in: [ID]
}

\\"\\"\\" @model \\"\\"\\"
type Note {
  id: ID!
  title: String!
  description: String!

  \\"\\"\\"@oneToMany field: 'note', key: 'test_id'\\"\\"\\"
  tests(filter: TestFilter): [Test]!
}

input NoteFilter {
  id: IDInput
  title: StringInput
  description: StringInput
  and: [NoteFilter]
  or: [NoteFilter]
  not: NoteFilter
}

type NoteResultList {
  items: [Note]!
  offset: Int
  limit: Int
  count: Int
}

input OrderByInput {
  field: String!
  order: SortDirectionEnum = ASC
}

input PageRequest {
  limit: Int
  offset: Int
}

type Query {
  getNote(id: ID!): Note
  findNotes(filter: NoteFilter, page: PageRequest, orderBy: OrderByInput): NoteResultList!
  getTest(id: ID!): Test
  findTests(filter: TestFilter, page: PageRequest, orderBy: OrderByInput): TestResultList!
}

enum SortDirectionEnum {
  DESC
  ASC
}

input StringInput {
  ne: String
  eq: String
  le: String
  lt: String
  ge: String
  gt: String
  in: [String]
  contains: String
  startsWith: String
  endsWith: String
}

\\"\\"\\"@model\\"\\"\\"
type Test {
  id: ID!
  name: String

  \\"\\"\\"@manyToOne field: 'tests', key: 'test_id'\\"\\"\\"
  note: Note
}

input TestFilter {
  id: IDInput
  name: StringInput
  test_id: IDInput
  and: [TestFilter]
  or: [TestFilter]
  not: TestFilter
}

type TestResultList {
  items: [Test]!
  offset: Int
  limit: Int
  count: Int
}
"
`;

exports[`Test snapshot config gql 1`] = `
"directive @test(description: String) on FIELD_DEFINITION | ENUM_VALUE

\\"\\"\\" @model \\"\\"\\"
type Comment {
  id: ID!
  title: String!
  description: String!

  \\"\\"\\"@manyToOne field: 'comments', key: 'note_id'\\"\\"\\"
  noteComment: Note!
}

input CommentFilter {
  id: IDInput
  title: StringInput
  description: StringInput
  note_id: IDInput
  and: [CommentFilter]
  or: [CommentFilter]
  not: CommentFilter
}

type CommentResultList {
  items: [Comment]!
  offset: Int
  limit: Int
  count: Int
}

input CommentSubscriptionFilter {
  id: ID
  title: String
  description: String
}

input CreateCommentInput {
  id: ID
  title: String!
  description: String!
  note_id: ID!
}

input CreateIssueInput {
  id: ID
  category: IssueCategory
  createdAt: DateTime
}

input CreateNoteInput {
  id: ID
  title: String!
  description: String!
  test_id: ID
}

input CreateTestInput {
  id: ID
  name: String
}

type CustomType {
  issue1432: String
}

scalar DateTime

input DateTimeInput {
  ne: DateTime
  eq: DateTime
  le: DateTime
  lt: DateTime
  ge: DateTime
  gt: DateTime
  in: [DateTime]
  between: [DateTime]
}

input IDInput {
  ne: ID
  eq: ID
  le: ID
  lt: ID
  ge: ID
  gt: ID
  in: [ID]
}

\\"\\"\\"
Model added to test schema generation using enum. 
See issue https://github.com/aerogear/graphback/issues/1299

@model
\\"\\"\\"
type Issue {
  id: ID!
  category: IssueCategory
  createdAt: DateTime

  \\"\\"\\"
  Field with custom type added as a fix for https://github.com/aerogear/graphback/issues/1432
  This shows that a custom type can be added without breaking Graphback schema generation
  \\"\\"\\"
  custom: CustomType
}

enum IssueCategory {
  BUG
  QUESTION
  FEATURE_REQUEST
}

input IssueFilter {
  id: IDInput
  category: StringInput
  createdAt: DateTimeInput
  and: [IssueFilter]
  or: [IssueFilter]
  not: IssueFilter
}

type IssueResultList {
  items: [Issue]!
  offset: Int
  limit: Int
  count: Int
}

input IssueSubscriptionFilter {
  id: ID
  category: IssueCategory
  createdAt: DateTime
}

input MutateCommentInput {
  id: ID!
  title: String
  description: String
  note_id: ID
}

input MutateIssueInput {
  id: ID!
  category: IssueCategory
  createdAt: DateTime
}

input MutateNoteInput {
  id: ID!
  title: String
  description: String
  test_id: ID
}

input MutateTestInput {
  id: ID!
  name: String
}

type Mutation {
  likeNote(id: ID!): Note!
  createNote(input: CreateNoteInput!): Note!
  updateNote(input: MutateNoteInput!): Note!
  deleteNote(input: MutateNoteInput!): Note!
  createComment(input: CreateCommentInput!): Comment!
  updateComment(input: MutateCommentInput!): Comment!
  deleteComment(input: MutateCommentInput!): Comment!
  createTest(input: CreateTestInput!): Test!
  updateTest(input: MutateTestInput!): Test!
  deleteTest(input: MutateTestInput!): Test!
  createIssue(input: CreateIssueInput!): Issue!
  updateIssue(input: MutateIssueInput!): Issue!
  deleteIssue(input: MutateIssueInput!): Issue!
}

\\"\\"\\" @model \\"\\"\\"
type Note {
  id: ID!
  title: String! @test(description: \\"mock directive\\")
  description: String!

  \\"\\"\\"@oneToMany field: 'noteComment', key: 'note_id'\\"\\"\\"
  comments(filter: CommentFilter): [Comment]!

  \\"\\"\\"@oneToOne key: 'test_id'\\"\\"\\"
  test: Test
}

input NoteFilter {
  id: IDInput
  title: StringInput
  description: StringInput
  test_id: IDInput
  and: [NoteFilter]
  or: [NoteFilter]
  not: NoteFilter
}

type NoteResultList {
  items: [Note]!
  offset: Int
  limit: Int
  count: Int
}

input NoteSubscriptionFilter {
  id: ID
  title: String
  description: String
}

input OrderByInput {
  field: String!
  order: SortDirectionEnum = ASC
}

input PageRequest {
  limit: Int
  offset: Int
}

type Query {
  getLikedNotes(id: ID!, names: [String]!): Note!
  getNote(id: ID!): Note
  findNotes(filter: NoteFilter, page: PageRequest, orderBy: OrderByInput): NoteResultList!
  getComment(id: ID!): Comment
  findComments(filter: CommentFilter, page: PageRequest, orderBy: OrderByInput): CommentResultList!
  getTest(id: ID!): Test
  findTests(filter: TestFilter, page: PageRequest, orderBy: OrderByInput): TestResultList!
  getIssue(id: ID!): Issue
  findIssues(filter: IssueFilter, page: PageRequest, orderBy: OrderByInput): IssueResultList!
}

enum SortDirectionEnum {
  DESC
  ASC
}

input StringInput {
  ne: String
  eq: String
  le: String
  lt: String
  ge: String
  gt: String
  in: [String]
  contains: String
  startsWith: String
  endsWith: String
}

type Subscription {
  newNote(filter: NoteSubscriptionFilter): Note!
  updatedNote(filter: NoteSubscriptionFilter): Note!
  deletedNote(filter: NoteSubscriptionFilter): Note!
  newComment(filter: CommentSubscriptionFilter): Comment!
  updatedComment(filter: CommentSubscriptionFilter): Comment!
  deletedComment(filter: CommentSubscriptionFilter): Comment!
  newTest(filter: TestSubscriptionFilter): Test!
  updatedTest(filter: TestSubscriptionFilter): Test!
  deletedTest(filter: TestSubscriptionFilter): Test!
  newIssue(filter: IssueSubscriptionFilter): Issue!
  updatedIssue(filter: IssueSubscriptionFilter): Issue!
  deletedIssue(filter: IssueSubscriptionFilter): Issue!
}

\\"\\"\\"@model\\"\\"\\"
type Test {
  id: ID!
  name: String
}

input TestFilter {
  id: IDInput
  name: StringInput
  and: [TestFilter]
  or: [TestFilter]
  not: TestFilter
}

type TestResultList {
  items: [Test]!
  offset: Int
  limit: Int
  count: Int
}

input TestSubscriptionFilter {
  id: ID
  name: String
}"
`;

exports[`Test snapshot config js 1`] = `
"directive @test(description: String) on FIELD_DEFINITION | ENUM_VALUE

\\"\\"\\" @model \\"\\"\\"
type Comment {
  id: ID!
  title: String!
  description: String!

  \\"\\"\\"@manyToOne field: 'comments', key: 'note_id'\\"\\"\\"
  noteComment: Note!
}

input CommentFilter {
  id: IDInput
  title: StringInput
  description: StringInput
  note_id: IDInput
  and: [CommentFilter]
  or: [CommentFilter]
  not: CommentFilter
}

type CommentResultList {
  items: [Comment]!
  offset: Int
  limit: Int
  count: Int
}

input CommentSubscriptionFilter {
  id: ID
  title: String
  description: String
}

input CreateCommentInput {
  id: ID
  title: String!
  description: String!
  note_id: ID!
}

input CreateIssueInput {
  id: ID
  category: IssueCategory
  createdAt: DateTime
}

input CreateNoteInput {
  id: ID
  title: String!
  description: String!
  test_id: ID
}

input CreateTestInput {
  id: ID
  name: String
}

type CustomType {
  issue1432: String
}

scalar DateTime

input DateTimeInput {
  ne: DateTime
  eq: DateTime
  le: DateTime
  lt: DateTime
  ge: DateTime
  gt: DateTime
  in: [DateTime]
  between: [DateTime]
}

input IDInput {
  ne: ID
  eq: ID
  le: ID
  lt: ID
  ge: ID
  gt: ID
  in: [ID]
}

\\"\\"\\"
Model added to test schema generation using enum. 
See issue https://github.com/aerogear/graphback/issues/1299

@model
\\"\\"\\"
type Issue {
  id: ID!
  category: IssueCategory
  createdAt: DateTime

  \\"\\"\\"
  Field with custom type added as a fix for https://github.com/aerogear/graphback/issues/1432
  This shows that a custom type can be added without breaking Graphback schema generation
  \\"\\"\\"
  custom: CustomType
}

enum IssueCategory {
  BUG
  QUESTION
  FEATURE_REQUEST
}

input IssueFilter {
  id: IDInput
  category: StringInput
  createdAt: DateTimeInput
  and: [IssueFilter]
  or: [IssueFilter]
  not: IssueFilter
}

type IssueResultList {
  items: [Issue]!
  offset: Int
  limit: Int
  count: Int
}

input IssueSubscriptionFilter {
  id: ID
  category: IssueCategory
  createdAt: DateTime
}

input MutateCommentInput {
  id: ID!
  title: String
  description: String
  note_id: ID
}

input MutateIssueInput {
  id: ID!
  category: IssueCategory
  createdAt: DateTime
}

input MutateNoteInput {
  id: ID!
  title: String
  description: String
  test_id: ID
}

input MutateTestInput {
  id: ID!
  name: String
}

type Mutation {
  likeNote(id: ID!): Note!
  createNote(input: CreateNoteInput!): Note!
  updateNote(input: MutateNoteInput!): Note!
  deleteNote(input: MutateNoteInput!): Note!
  createComment(input: CreateCommentInput!): Comment!
  updateComment(input: MutateCommentInput!): Comment!
  deleteComment(input: MutateCommentInput!): Comment!
  createTest(input: CreateTestInput!): Test!
  updateTest(input: MutateTestInput!): Test!
  deleteTest(input: MutateTestInput!): Test!
  createIssue(input: CreateIssueInput!): Issue!
  updateIssue(input: MutateIssueInput!): Issue!
  deleteIssue(input: MutateIssueInput!): Issue!
}

\\"\\"\\" @model \\"\\"\\"
type Note {
  id: ID!
  title: String!
  description: String!

  \\"\\"\\"@oneToMany field: 'noteComment', key: 'note_id'\\"\\"\\"
  comments(filter: CommentFilter): [Comment]!

  \\"\\"\\"@oneToOne key: 'test_id'\\"\\"\\"
  test: Test
}

input NoteFilter {
  id: IDInput
  title: StringInput
  description: StringInput
  test_id: IDInput
  and: [NoteFilter]
  or: [NoteFilter]
  not: NoteFilter
}

type NoteResultList {
  items: [Note]!
  offset: Int
  limit: Int
  count: Int
}

input NoteSubscriptionFilter {
  id: ID
  title: String
  description: String
}

input OrderByInput {
  field: String!
  order: SortDirectionEnum = ASC
}

input PageRequest {
  limit: Int
  offset: Int
}

type Query {
  getLikedNotes(id: ID!, names: [String]!): Note!
  getNote(id: ID!): Note
  findNotes(filter: NoteFilter, page: PageRequest, orderBy: OrderByInput): NoteResultList!
  getComment(id: ID!): Comment
  findComments(filter: CommentFilter, page: PageRequest, orderBy: OrderByInput): CommentResultList!
  getTest(id: ID!): Test
  findTests(filter: TestFilter, page: PageRequest, orderBy: OrderByInput): TestResultList!
  getIssue(id: ID!): Issue
  findIssues(filter: IssueFilter, page: PageRequest, orderBy: OrderByInput): IssueResultList!
}

enum SortDirectionEnum {
  DESC
  ASC
}

input StringInput {
  ne: String
  eq: String
  le: String
  lt: String
  ge: String
  gt: String
  in: [String]
  contains: String
  startsWith: String
  endsWith: String
}

type Subscription {
  newNote(filter: NoteSubscriptionFilter): Note!
  updatedNote(filter: NoteSubscriptionFilter): Note!
  deletedNote(filter: NoteSubscriptionFilter): Note!
  newComment(filter: CommentSubscriptionFilter): Comment!
  updatedComment(filter: CommentSubscriptionFilter): Comment!
  deletedComment(filter: CommentSubscriptionFilter): Comment!
  newTest(filter: TestSubscriptionFilter): Test!
  updatedTest(filter: TestSubscriptionFilter): Test!
  deletedTest(filter: TestSubscriptionFilter): Test!
  newIssue(filter: IssueSubscriptionFilter): Issue!
  updatedIssue(filter: IssueSubscriptionFilter): Issue!
  deletedIssue(filter: IssueSubscriptionFilter): Issue!
}

\\"\\"\\"@model\\"\\"\\"
type Test {
  id: ID!
  name: String
}

input TestFilter {
  id: IDInput
  name: StringInput
  and: [TestFilter]
  or: [TestFilter]
  not: TestFilter
}

type TestResultList {
  items: [Test]!
  offset: Int
  limit: Int
  count: Int
}

input TestSubscriptionFilter {
  id: ID
  name: String
}
"
`;

exports[`Test snapshot config ts 1`] = `
"directive @test(description: String) on FIELD_DEFINITION | ENUM_VALUE

\\"\\"\\" @model \\"\\"\\"
type Comment {
  id: ID!
  title: String!
  description: String!

  \\"\\"\\"@manyToOne field: 'comments', key: 'note_id'\\"\\"\\"
  noteComment: Note!
}

input CommentFilter {
  id: IDInput
  title: StringInput
  description: StringInput
  note_id: IDInput
  and: [CommentFilter]
  or: [CommentFilter]
  not: CommentFilter
}

type CommentResultList {
  items: [Comment]!
  offset: Int
  limit: Int
  count: Int
}

input CommentSubscriptionFilter {
  id: ID
  title: String
  description: String
}

input CreateCommentInput {
  id: ID
  title: String!
  description: String!
  note_id: ID!
}

input CreateIssueInput {
  id: ID
  category: IssueCategory
  createdAt: DateTime
}

input CreateNoteInput {
  id: ID
  title: String!
  description: String!
  test_id: ID
}

input CreateTestInput {
  id: ID
  name: String
}

type CustomType {
  issue1432: String
}

scalar DateTime

input DateTimeInput {
  ne: DateTime
  eq: DateTime
  le: DateTime
  lt: DateTime
  ge: DateTime
  gt: DateTime
  in: [DateTime]
  between: [DateTime]
}

input IDInput {
  ne: ID
  eq: ID
  le: ID
  lt: ID
  ge: ID
  gt: ID
  in: [ID]
}

\\"\\"\\"
Model added to test schema generation using enum. 
See issue https://github.com/aerogear/graphback/issues/1299

@model
\\"\\"\\"
type Issue {
  id: ID!
  category: IssueCategory
  createdAt: DateTime

  \\"\\"\\"
  Field with custom type added as a fix for https://github.com/aerogear/graphback/issues/1432
  This shows that a custom type can be added without breaking Graphback schema generation
  \\"\\"\\"
  custom: CustomType
}

enum IssueCategory {
  BUG
  QUESTION
  FEATURE_REQUEST
}

input IssueFilter {
  id: IDInput
  category: StringInput
  createdAt: DateTimeInput
  and: [IssueFilter]
  or: [IssueFilter]
  not: IssueFilter
}

type IssueResultList {
  items: [Issue]!
  offset: Int
  limit: Int
  count: Int
}

input IssueSubscriptionFilter {
  id: ID
  category: IssueCategory
  createdAt: DateTime
}

input MutateCommentInput {
  id: ID!
  title: String
  description: String
  note_id: ID
}

input MutateIssueInput {
  id: ID!
  category: IssueCategory
  createdAt: DateTime
}

input MutateNoteInput {
  id: ID!
  title: String
  description: String
  test_id: ID
}

input MutateTestInput {
  id: ID!
  name: String
}

type Mutation {
  likeNote(id: ID!): Note!
  createNote(input: CreateNoteInput!): Note!
  updateNote(input: MutateNoteInput!): Note!
  deleteNote(input: MutateNoteInput!): Note!
  createComment(input: CreateCommentInput!): Comment!
  updateComment(input: MutateCommentInput!): Comment!
  deleteComment(input: MutateCommentInput!): Comment!
  createTest(input: CreateTestInput!): Test!
  updateTest(input: MutateTestInput!): Test!
  deleteTest(input: MutateTestInput!): Test!
  createIssue(input: CreateIssueInput!): Issue!
  updateIssue(input: MutateIssueInput!): Issue!
  deleteIssue(input: MutateIssueInput!): Issue!
}

\\"\\"\\" @model \\"\\"\\"
type Note {
  id: ID!
  title: String!
  description: String!

  \\"\\"\\"@oneToMany field: 'noteComment', key: 'note_id'\\"\\"\\"
  comments(filter: CommentFilter): [Comment]!

  \\"\\"\\"@oneToOne key: 'test_id'\\"\\"\\"
  test: Test
}

input NoteFilter {
  id: IDInput
  title: StringInput
  description: StringInput
  test_id: IDInput
  and: [NoteFilter]
  or: [NoteFilter]
  not: NoteFilter
}

type NoteResultList {
  items: [Note]!
  offset: Int
  limit: Int
  count: Int
}

input NoteSubscriptionFilter {
  id: ID
  title: String
  description: String
}

input OrderByInput {
  field: String!
  order: SortDirectionEnum = ASC
}

input PageRequest {
  limit: Int
  offset: Int
}

type Query {
  getLikedNotes(id: ID!, names: [String]!): Note!
  getNote(id: ID!): Note
  findNotes(filter: NoteFilter, page: PageRequest, orderBy: OrderByInput): NoteResultList!
  getComment(id: ID!): Comment
  findComments(filter: CommentFilter, page: PageRequest, orderBy: OrderByInput): CommentResultList!
  getTest(id: ID!): Test
  findTests(filter: TestFilter, page: PageRequest, orderBy: OrderByInput): TestResultList!
  getIssue(id: ID!): Issue
  findIssues(filter: IssueFilter, page: PageRequest, orderBy: OrderByInput): IssueResultList!
}

enum SortDirectionEnum {
  DESC
  ASC
}

input StringInput {
  ne: String
  eq: String
  le: String
  lt: String
  ge: String
  gt: String
  in: [String]
  contains: String
  startsWith: String
  endsWith: String
}

type Subscription {
  newNote(filter: NoteSubscriptionFilter): Note!
  updatedNote(filter: NoteSubscriptionFilter): Note!
  deletedNote(filter: NoteSubscriptionFilter): Note!
  newComment(filter: CommentSubscriptionFilter): Comment!
  updatedComment(filter: CommentSubscriptionFilter): Comment!
  deletedComment(filter: CommentSubscriptionFilter): Comment!
  newTest(filter: TestSubscriptionFilter): Test!
  updatedTest(filter: TestSubscriptionFilter): Test!
  deletedTest(filter: TestSubscriptionFilter): Test!
  newIssue(filter: IssueSubscriptionFilter): Issue!
  updatedIssue(filter: IssueSubscriptionFilter): Issue!
  deletedIssue(filter: IssueSubscriptionFilter): Issue!
}

\\"\\"\\"@model\\"\\"\\"
type Test {
  id: ID!
  name: String
}

input TestFilter {
  id: IDInput
  name: StringInput
  and: [TestFilter]
  or: [TestFilter]
  not: TestFilter
}

type TestResultList {
  items: [Test]!
  offset: Int
  limit: Int
  count: Int
}

input TestSubscriptionFilter {
  id: ID
  name: String
}
"
`;
